### PR TITLE
EGM-certified electron IDs for 2016 data: cut-based and MVA

### PR DESCRIPTION
**Pull request by ikrav**
This PR contains EGM certified cut-based and MVA electron IDs tuned on 80X samples. They are intended to be used on 2016 data. The electron MVA contains the HZZ MVA used for ICHEP16 and the presently recommended general purpose MVA tuned for better performance given features of a more complete 2016 dataset.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2Spring16GeneralPurpose.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+#include "DataFormats/EgammaCandidates/interface/ConversionFwd.h"
+#include "DataFormats/EgammaCandidates/interface/Conversion.h"
+#include "RecoEgamma/EgammaTools/interface/ConversionTools.h"
+
+#include "CondFormats/EgammaObjects/interface/GBRForest.h"
+
+#include <vector>
+#include <string>
+#include <TROOT.h>
```
Comment (slava77) [inline RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2Spring16GeneralPurpose.h]: is TROOT include really needed?

**Comment by cvuosalo** [comment]
This PR needs to be re-based.

**Comment by ikrav** [comment]
TROOT.h is included in all past MVA headers of this type for electron/photon IDs. 
It was included from the start because TMVA tutorials always include it. I do not know if it
is totally necessary to include it, I guess I can check and see if it works without it.

**Comment by ikrav** [comment]
Carl, does this mean I need to do something? If yes, can you give me
guidance?

**Comment by slava77** [comment]
@ikrav 
from tomorrow you should be able to test and submit a pull request in 80X.
The data files should make it to the new IB (probably CMSSW_8_0_X-12-07-2300)

For rebase instructions you may want to check 
https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideHATS2016Git#CMSSW_rebasing_recipe

Your proposed code change is currently in conflict with  RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.h  file in 90X release (as noted at the bottom of this PR thread web page https://github.com/cms-sw/cmssw/pull/16839 )

**Comment by ikrav** [comment]
What should I rebase? The PR I submitted to 90X? Or the PR I am about
to submit to 80X? All of them? Is this a rule, do I need to rebase anything to
the release to which I am submitting the PR before I submit the PR?
-thanks

**Comment by slava77** [comment]
On 12/7/16 12:12 PM, ikrav wrote:
> What should I rebase? The PR I submitted to 90X? Or the PR I am about
> to submit to 80X? All of them? Is this a rule, do I need to rebase
> anything to
> the release to which I am submitting the PR before I submit the PR?
> -thanks


You need to rebase the PR that exists and is said to have a merge
conflict (this 90X PR).

Ideally, you need to prepare the PR topic branch in the same release
that you submit it to.

**Comment by ikrav** [comment]
Ok, I will rebase the 90X PR. Sorry for many questions, but what is the merge conflict
you see? I merged my private branch into 90X IB most recent release when I was submitting,
and it merged fine and built fine.

**Comment by slava77** [comment]
On 12/7/16 12:25 PM, ikrav wrote:
> Ok, I will rebase the 90X PR. Sorry for many questions, but what is the
> merge conflict
> you see? I merged my private branch into 90X IB most recent release when
> I was submitting,
> and it merged fine and built fine.

I did not look which lines actually conflict.
There is a message from automated checking
 "This branch has conflicts that must be resolved"
I hope you can see it.

File: RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Spring16GeneralPurpose.cc [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+#include "DataFormats/PatCandidates/interface/Electron.h"
+
+#include "FWCore/ParameterSet/interface/FileInPath.h"
+
+#include "TMath.h"
+#include "TMVA/MethodBDT.h"
+
+ElectronMVAEstimatorRun2Spring16GeneralPurpose::ElectronMVAEstimatorRun2Spring16GeneralPurpose(const edm::ParameterSet& conf):
+  AnyMVAEstimatorRun2Base(conf),
+  _tag(conf.getParameter<std::string>("mvaTag")),
```
Comment (cvuosalo) [inline RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Spring16GeneralPurpose.cc]: Variable names starting with "_" violate the CMS coding rules. They need to be changed.

**Comment by ikrav** [comment]
> On Dec 7, 2016, at 2:37 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> I did not look which lines actually conflict.
> There is a message from automated checking
> "This branch has conflicts that must be resolved"
> I hope you can see it.

I do see now that:

This branch has conflicts that must be resolved
Only those with write access to this repository can merge pull requests.
Conflicting files
RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.h

But how can I learned more about it? I just checked out a fresh CMSSW_9_0_X_2016-12-07-0000.
I merged into it my branch ikrav:egm_id_80X_v2 which I used for this PR. I see no merging
conflicts. The GitHub page of the PR tells me no information beyond the name of the
file above. How can I find what and where is wrong?

File: RecoEgamma/ElectronIdentification/python/ElectronMVAValueMapProducer_cfi.py [hunk 3]
```diff
@@ -15,6 +15,13 @@
 from RecoEgamma.ElectronIdentification.Identification.mvaElectronID_Spring15_25ns_Trig_V1_cff import *
 mvaConfigsForEleProducer.append( mvaEleID_Spring15_25ns_Trig_V1_producer_config )
 
+from RecoEgamma.ElectronIdentification.Identification.mvaElectronID_Spring16_HZZ_V1_cff import *
```
Comment (cvuosalo) [inline RecoEgamma/ElectronIdentification/python/ElectronMVAValueMapProducer_cfi.py]: Please don't use `import *`. Please list the modules you need to import. There are many instances of `import *` in this PR that should be fixed.

**Comment by davidlange6** [comment]
i don’t see a big need to avoid import * - standard python practice. Is this cff particularly big?

**Comment by ikrav** [comment]
Ok, I followed instructions for rebasing and created a new branch in my-cmssw
based on a recent 90X release. Now the question is what to do with it. I am reluctant
to update the original branch that was used to submit the PR (based on 8_0_10) 
with the rebased version because it is already publicized in recipes of electron IDs. 
Any advice?

**Comment by cvuosalo** [comment]
Reco would prefer to reduce the overhead of importing and constructing large numbers of unused modules. If the `import *` only imports modules that are truly needed, then it might be OK, but later additions to the imported file could then be imported unnecessarily.

**Comment by cvuosalo** [comment]
A test of workflow 25202.0_TTbar_13 with no DQM or Validation with 150 events against baseline CMSSW_9_0_0_pre1 shows no significant differences and no increase in memory usage. CPU time increases very marginally:
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   +0.478014      +0.01%         0.62 ms/ev ->         1.00 ms/ev electronMVAValueMapProducer
  ---------- ------------     --------                  ----       ------------
```

**Comment by ikrav** [comment]
I replaced several import * by explicit lists (only the new ones, though. I have not
changed those import * that are already in CMSSW). Additionally, I removed TROOT.h
includes, another request. Finally, names with leading underscores are renamed. 
All of this is in my new rebased to 90X branch named differently from the branch that
started this PR. Now my question is how to proceed (same question was also in another
comment from me from a few hours ago).

**Comment by cvuosalo** [comment]
@ikrav: If you're still having problems with your new branch, you could just create a new PR and then close this one.

**Comment by cvuosalo** [comment]
@ikrav: If you want this PR in 900pre2, it needs to be fixed today (Friday).

**Comment by ikrav** [comment]
Carl, I do not have problems with my new branch. Rather, I do not know what to do 
with it. I have the branch that I used to create the PR. I do not want to change
that branch if possible because it is already being used by people in recipes.
I created a new branch with changes you and Slava requested. Now my question
is whether somehow this PR can use the new branch, or whether I have to cancel
this one and start new one, or whether a new PR can be started from this PR if such
a thing is possible, etc. I do not have experience, so I do not know what my steps are,
and it seems that there are not too many directly involved experts in my time zone.

**Comment by cvuosalo** [comment]
@ikrav: Creating a new PR would seem the simplest approach. I'm not understanding why that hasn't been done yet.

**Comment by ikrav** [comment]
Carl, this has not been done because I was waiting for explicit guidance from you
or Slava. I will create a new PR today.

**Comment by ikrav** [comment]
This PR is closed as I just submitted a new PR from a rebased to 90X branch. The new PR number is #16979.

**Pull request closed without merging**